src/SemaphoreAirdrops.sol

```diff
@@ -0,0 +1,149 @@
+    ///                                  EVENTS                                ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Emitted when an airdrop is created
+    /// @param airdropId The id of the airdrop
+    /// @param airdrop The airdrop details
+    event AirdropCreated(uint256 airdropId, Airdrop airdrop);
+
+    /// @notice Emitted when an airdrop is successfully claimed
+    /// @param receiver The address that received the airdrop
+    event AirdropClaimed(uint256 indexed airdropId, address receiver);
+
+    /// @notice Emitted when the airdropped amount is changed
+    /// @param airdropId The id of the airdrop getting updated
+    /// @param airdrop The new details for the airdrop
+    event AirdropUpdated(uint256 indexed airdropId, Airdrop airdrop);
+
+    struct Airdrop {
+        uint256 groupId;
+        ERC20 token;
+        address manager;
+        address holder;
+        uint256 amount;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                              CONFIG STORAGE                            ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @dev The Semaphore instance that will be used for managing groups and verifying proofs
+    ISemaphore internal immutable semaphore;
+
+    /// @dev Wether a nullifier hash has been used already. Used to prevent double-signaling
+    mapping(uint256 => bool) internal nullifierHashes;
+
+    uint256 internal airdropId = 1;
+    mapping(uint256 => Airdrop) public getAirdrop;
+
+    ///////////////////////////////////////////////////////////////////////////////
+    ///                               CONSTRUCTOR                              ///
+    //////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Deploys a SemaphoreAirdrop instance
+    /// @param _semaphore The Semaphore instance that will manage groups and verify proofs
+    constructor(
+        ISemaphore _semaphore
+    ) {
+        semaphore = _semaphore;
+    }
+
+    /// @notice Create a new airdrop
+    /// @param groupId The ID of the Semaphore group that will be elegible to claim this airdrop
+    /// @param token The ERC20 token that will be airdropped to elegible participants
+    /// @param holder The address holding the tokens that will be airdropped
+    /// @param amount The amount of tokens that each participant will receive upon claiming
+    function createAirdrop(
+        uint256 groupId,
+        ERC20 token,
+        address holder,
+        uint256 amount
+    ) public {
+        Airdrop memory airdrop = Airdrop({
+            groupId: groupId,
+            token: token,
+            manager: msg.sender,
+            holder: holder,
+            amount: amount
+        });
+
+        getAirdrop[airdropId] = airdrop;
+        emit AirdropCreated(airdropId++, airdrop);
```

Comment:
Having side-effects  argument positions is dangerous. For example

```
int counter = 0;
emit MyEvent(counter, counter++);
```

can result in either `MyEvent(0, 0)` or `MyEvent(1, 0)`. Solidity docs warns that it is undefined what it does. 

https://docs.soliditylang.org/en/v0.8.13/control-structures.html#order-of-evaluation-of-expressions

in fact, Solidity changing what it does was a a key part of the winning entry in the underhanded Solidity contest:

https://blog.soliditylang.org/2022/04/09/announcing-the-underhanded-contest-winners-2022/

Reply:
Right, should be fine in this case since we're not reusing `airdropId` after incrementing tho right?

Reply:
Correct. It's more of a code-style thing. Like how you would generally do external calls after state changes, even if there is no risk of a re-entrancy attacks.